1-Draft/RFCXXXX-Obsoletion-and-Deprecation.md

```diff
@@ -0,0 +1,511 @@
+>> param ( )
+>> }
+PS> Invoke-Obsolete1
+WARNING: The command 'Invoke-Obsolete1' is obsolete. Invoke-Obsolete1 is no longer supported
+```
+
+Ex.2 Obsolete parameter when the parameter is used
+
+```powershell
+PS> function Invoke-Obsolete2 {
+>> param (
+>> [ObsoleteAttribute("Use parameter2")]
+>> [string]$parameter1,
+>> [string]$parameter2
+>> )
+>> }
+PS> Invoke-Obsolete2 -parameter1 value1
+WARNING: Parameter 'parameter1' is obsolete. Use parameter2
+```
+
+Ex.3 Obsolete cmdlet and parameter are both used
+
+```powershell
+PS> function Invoke-Obsolete3 {
+>> [ObsoleteAttribute("Use a different cmdlet")]
+>> param (
+>> [ObsoleteAttribute("Use parameter2")]
+>> [string]$parameter1,
+>> [string]$parameter2
+>> )
+>> }
+PS> Invoke-Obsolete3 -parameter1 value1
+WARNING: The command 'Invoke-Obsolete3' is obsolete. Use a different cmdlet
+WARNING: Parameter 'parameter1' is obsolete. Use parameter2
+```
+
+Ex.4 obsolete parameter is not used
+
+```powershell
+
+PS> function Invoke-Obsolete4 {
+>> param (
+>> [ObsoleteAttribute("Use parameter2")]
+>> [string]$parameter1,
+>> [string]$parameter2
+>> )
+>> }
+PS> Invoke-Obsolete3 -parameter2 value2
+```
+
+Note that the warning is only emitted when the obsolete parameter _is used_.
+If you use a parameter which is not marked obsolete, naturally no warning will be issued.
+
+### Interacting with Types
+
+The cmdlet behavior is not available for types, as the following example shows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Management.Automation" Version="7.0.3" />
+  </ItemGroup>
+</Project>
+```
+
+```cs
+using System;
+using System.Management.Automation;
+
+namespace p1
+{
+    public class Class1
+    {
+        public PowerShellStreamType streamtype;
+    }
+}
+```
+
+```powershell
+PS> dotnet build
+Microsoft (R) Build Engine version 16.8.0-preview-20414-02+a55ce4fbb for .NET
+Copyright (C) Microsoft Corporation. All rights reserved.
+
+  Determining projects to restore...
+  All projects are up-to-date for restore.
+  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
+/Users/james/src/projects/obsolete/p1/Class1.cs(8,16): error CS0619: 'PowerShellStreamType' is obsolete: 'This enum type was used only in PowerShell Workflow and is now obsolete.' [/Users/james/src/projects/obsolete/p1/p1.csproj]
+
+Build FAILED.
+
+/Users/james/src/projects/obsolete/p1/Class1.cs(8,16): error CS0619: 'PowerShellStreamType' is obsolete: 'This enum type was used only in PowerShell Workflow and is now obsolete.' [/Users/james/src/projects/obsolete/p1/p1.csproj]
+    0 Warning(s)
+    1 Error(s)
+
+Time Elapsed 00:00:01.07
+```
+
+However, I can use this type from PowerShell without warning or error
+
+```powershell
+ PS> $a = [System.Management.Automation.PowerShellStreamType]::Information.Value__
```

Comment:
```suggestion
 PS> $a = [System.Management.Automation.PowerShellStreamType]::Information
```

The name of an enum's underlying field isn't strictly required to be `value__` by the runtime, only by CLS compliance rules.  Though it's unlikely a compiler will ever use a different name, it's direct usage probably shouldn't be encouraged.